src/file_pipe_log/pipe.rs

```diff
@@ -147,19 +149,34 @@ impl<F: FileSystem> SinglePipe<F> {
 
     /// Synchronizes all metadatas associated with the working directory to the
     /// filesystem.
-    fn sync_dir(&self) -> Result<()> {
+    fn sync_dir(&self, path_id: PathId) -> Result<()> {
         debug_assert!(!self.paths.is_empty());
-        let path = PathBuf::from(&self.paths[0]);
-        std::fs::File::open(path).and_then(|d| d.sync_all())?;
+        std::fs::File::open(&PathBuf::from(&self.paths[path_id])).and_then(|d| d.sync_all())?;
         Ok(())
     }
 
-    fn new_file(&self, seq: FileSeq) -> Result<(PathId, F::Handle)> {
+    fn new_file(&self, seq: FileSeq, expected_file_size: usize) -> Result<(PathId, F::Handle)> {
         let new_file_id = FileId {
             seq,
             queue: self.queue,
         };
-        if let Some(f) = self.recycled_files.write().pop_front() {
+        if expected_file_size > self.target_file_size {
+            // On this state, it means that there already exists `nospace` errors and
+            // current buffer is too huge to be appended directly. So, we should free
+            // enough space by clearing several recycled log files until this buffer
+            // could be dumped.
+            let mut expected_size = expected_file_size;
```

Comment:
(1) recycled file could be on a different disk than the full one
(2) should not modify the `expected_file_size` passed into `fetch_dir`. There could be another writer occupying the space you just released.
(3) if deleting recycled file is enough, there's no need to rotate, which makes the rotated file smaller than usual and hurt future performance.

This branch needs test.

Reply:
Yep, these cases have already considered in this branch.

Firstly, about:
>(1) recycled file could be on a different disk than the full one

That's right, this case has three corner cases(Suppose `main-dir` is full, and if and only if the first attempt by `engine.write(...)` is failed, will the whole writing progress try to enter this branch.):
+ 1. `main-dir` has no recycled logs, and `spill-dir` have several recycled logs. 
As the `main-dir` is full, the following `find_available_dir` must skip the `main-dir`, and `spill-dir` should prepare enough space for this write. Otherwise, the whole write progress(`engine.write(...)`) will fail finally with `Err:TryAgain`.
So, in this case, the deletion is necessary for preparing enough space for write. And if there did not exists enough space for writing this `LogBatch`, it will fail as expected because there truly exists no enough space on both disks.
+ 2. `main-dir` has several recycled logs, and `sum(recycled_logs.size())` < `expected_file_size`, but no recycled logs exist in `spill-dir`.
After deletion, the `main-dir` still have no enough space for this second attempt. So, the following `find_available_dir` must skip the `main-dir` as expected. And whether this second attempt on write will success or not depends on whether there exists enough space on `spill-dir`. It's expected. So, int this case, this deletion is also necessary.
+ 3. Both `main-dir` and `spill-dir` have several recycled logs.
The deletion progress will try to release free space on one dir until it's enough to dump this LogBatch into this dir. If and only if free space is still not enough, that is, `deleted_recycled_logs.size() + free_space.size()` < `expected_file_size`, does fail this second write. So, the deletion should be kept.

Secondly, about:
> (2) should not modify the expected_file_size passed into fetch_dir. There could be another writer occupying the space you just released.

Here, the `expected_file_size` do not be modified, just copy it to `expected_size` in this pr.

And for:
> (3) if deleting recycled file is enough, there's no need to rotate, which makes the rotated file smaller than usual and hurt future performance.

Yep, unless `expected_file_size` <= `self.target_file_size`, it will do reuse recycled logs by `rotate`. Otherwise, it will try to free enough space and find an available dir to place this LogBatch by the following `find_available_dir` directly.

As for testing on this branch, I've do it by the following manual case on `centos7`:
```Rust
#[test]
fn test_build_engine_with_spill_dir() {
    fn mount_tmpfs(path: &str) {
        debug_assert!(!path.is_empty());
        #[cfg(target_os = "linux")]
        {
            use std::process::Command;

            let mut mk_device = Command::new("sudo");
            mk_device.args([
                "mount",
                "-t",
                "tmpfs",
                "-o",
                "size=128M",
                "tmpfs",
                path.into(),
            ]);
            mk_device.status().expect("process failed to execute");
        }
    }

    fn umount_tmpfs(path: &str) {
        debug_assert!(!path.is_empty());
        #[cfg(target_os = "linux")]
        {
            use std::process::Command;

            let mut mk_device = Command::new("sudo");
            mk_device.args(["umount", path.into()]);
            mk_device.status().expect("process failed to execute");
        }
    }

    let main_dir = tempfile::Builder::new()
        .prefix("test_build_engine_with_spill_dir_main")
        .tempdir()
        .unwrap();
    let spill_dir = tempfile::Builder::new()
        .prefix("test_build_engine_with_spill_dir_spill")
        .tempdir()
        .unwrap();
    {
        // Mount a tmpfs with limited size.
        mount_tmpfs(main_dir.path().to_str().unwrap());
    }
    let cfg = Config {
        dir: main_dir.path().to_str().unwrap().to_owned(),
        spill_dir: Some(spill_dir.path().to_str().unwrap().to_owned()),
        target_file_size: ReadableSize::mb(1),
        purge_threshold: ReadableSize::mb(128),
        enable_log_recycle: true,
        prefill_for_recycle: true,
        ..Default::default()
    };
    let data = vec![b'x'; 17371];
    let engine = Engine::open(cfg.clone()).unwrap();
    let _f1 = FailGuard::new("file_pipe_log::force_choose_dir", "return");
    for rid in 1..12000 {
        assert!(catch_unwind_silent(|| append(&engine, rid, 1, 100, Some(&data))).is_ok());
    }
    {
        // TODO: failed because current pid still occupy this dir.
        umount_tmpfs(main_dir.path().to_str().unwrap());
    }
}
```
But I'm wondering how to release the temporarily mounted device and make it approved by CI/CD.

Reply:
Honestly, I don't even know how to respond to such a long comment. If the situation is indeed so complicated, you should comment it in the code so that we can discuss line by line there.

> (1.i) So, in this case, the deletion is necessary for preparing enough space for write.

You didn't check the available space of spill_dir. It might not need deletion at all.

> (1.iii) The deletion progress will try to release free space on one dir until it's enough to dump this LogBatch into this dir.

Same thing as before, you don't know if the other directory has enough available space so that it can avoid deletion. And in addition to that, your code won't work if the sum of deleted files from dir1 and dir2 reaches target_size, but the the two individual directories still haven't deleted enough.

> (2) Here, the expected_file_size do not be modified, just copy it to expected_size in this pr.

Well, it's a misread, but you shouldn't copy something into another with nearly the same name.

> Yep, unless expected_file_size <= self.target_file_size, it will do reuse recycled logs by rotate. Otherwise, it will try to free enough space and find an available dir to place this LogBatch by the following find_available_dir directly.

I think there's some misunderstanding. I meant you don't need to rotate the active file, i.e. no need to close the file writer and open a new one.

Reply:
> As for testing on this branch, I've do it by the following manual case on centos7:

Just some failpoint unit tests are enough. Hardware doesn't really matter to this particular function. I only reason I asked before to test on hardware, is because I don't know if truncating a file on an already full disk can succeed.

Reply:
> You didn't check the available space of spill_dir. It might not need deletion at all.

IMO, the too early and unified `deletion` is focused on reducing the extra checking of `available_space` on each dir, and unify the code into one branch `expected_file_size > target_file_size`. For example, the `available_space` consists of recycled logs and free space on each dir respectively, and the checking and deletion of recycled files should be separated individually.

As this pr has paid too long comments on this point, making it too ugly to be reviewed, I accept the [previous advice](https://github.com/tikv/raft-engine/pull/294#discussion_r1121174848) that we should open another issue to focus on optimizing this issue.